### PR TITLE
[PLAT-153] [mds-compliance-engine] [mds-policy-service] Improve Compliance time checking & Policy time validation

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
This PR fixes a bug in the compliance engine where time-of-day checks were not working properly. Previously with Moment.js (note: a deprecated library), we were creating Moment objects from the time strings given to us within the Policy _without_ specifying timezone. As part of the refactor to switch to Luxon, I discovered that additionally our logic for checking boundaries (specifically across days, or with 'overnight' policies) was lackluster as well. And then, I realized that our validation for `start_time`/`end_time` was also pretty bad.

This PR switches Compliance to Luxon instead of Moment for nearly everything (besides TIMEZONE env var validation, still need to use Moment for that), and applies stricter regex validation for `start_time` and `end_time` when creating Policies.

## 👌 Resolves:
- Fixes Compliance time-of-day checks
- Adds stricter validation to the Policy Service

## 📦 Impacts:
See changesets.

## ✅ PR Checklist
- [x] Write an incredible amount of tests

### CODE REVIEW

File: packages/mds-compliance-engine/engine/helpers.ts [hunk 1]
```diff
@@ -98,20 +184,15 @@ export function isRuleActive(rule: Rule<Exclude<RULE_TYPE, 'rate'>>): boolean {
...
-      }
-    }
-  }
-  return false
+  return isCurrentDayInDays({ days }) && isCurrentTimeInInterval({ start_time, end_time })
 }
 
 export function isInVehicleTypes(rule: Rule<Exclude<RULE_TYPE, 'rate'>>, device: Device): boolean {
-  return !rule.vehicle_types || (rule.vehicle_types && rule.vehicle_types.includes(device.vehicle_type))
+  return (
```
Comment (avatarneil) [inline packages/mds-compliance-engine/engine/helpers.ts:191]: Oh yeah, this was another fun bug I found :)

**Review by nessur — APPROVED** [review]

File: packages/mds-compliance-engine/tests/engine/helpers.spec.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+          days: null,
+          start_time: '09:00:00',
+          end_time: '10:00:00'
+        }),
+        false
+      )
+    })
+
+    describe('Tests edge case: start_time > end_time (across midnight). e.g. 7pm-5am', () => {
+      it('Tests before midnight (true)', () => {
```
Comment (mgeorgehansen-lacuna) [inline packages/mds-compliance-engine/tests/engine/helpers.spec.ts:85]: May want to add two test cases for current time past midnight (early morning), e.g.:

1. current time is 04:00:00 (should still be active since start_date >= end_date)
2. current time is 06:00:00 (should no longer be active, past end_time)

**Review by mgeorgehansen-lacuna — APPROVED** [review]
I think this looks solid. There's a possible edge case when the start time of a rule falls within a DST transition (hour is repeated), but it's probably not worth fixing IMO.

File: packages/mds-compliance-engine/engine/helpers.ts [hunk 3]
```diff
@@ -69,7 +80,66 @@ export function isPolicyActive(policy: PolicyDomainModel, end_time: number = now
...
+ * @returns If the current local time is within the specified time range.
+ *
+ * Note: We can just do string comparison here for times, because the format should consistently be HH:mm:ss for all strings.
+ */
+const isCurrentTimeInInterval = ({ start_time, end_time }: Pick<Rule, 'start_time' | 'end_time'>) => {
+  const { TIMEZONE } = env
+
+  const currentTime = DateTime.now().setZone(TIMEZONE)
+  const formattedCurrentTime = currentTime.toFormat(TIME_FORMAT)
+  if (!isOfTimeFormat(formattedCurrentTime))
```
Comment (janedotx) [inline packages/mds-compliance-engine/engine/helpers.ts:124]: Is this really necessary? You just formatted the current time yourself using an internally determined format, so how would this ever throw?

**Review by janedotx — APPROVED** [review]

File: packages/mds-compliance-engine/engine/helpers.ts [hunk 4]
```diff
@@ -78,20 +148,15 @@ export function isRuleActive(rule: Rule<Exclude<RULE_TYPE, 'rate'>>): boolean {
...
-  return false
+  return isCurrentDayInDays({ days }) && isCurrentTimeInInterval({ start_time, end_time })
 }
 
 export function isInVehicleTypes(rule: Rule<Exclude<RULE_TYPE, 'rate'>>, device: Device): boolean {
-  return !rule.vehicle_types || (rule.vehicle_types && rule.vehicle_types.includes(device.vehicle_type))
+  return (
+    !rule.vehicle_types ||
+    rule.vehicle_types.length === 0 ||
+    (rule.vehicle_types && rule.vehicle_types.includes(device.vehicle_type))
```
Comment (PeterSP) [inline packages/mds-compliance-engine/engine/helpers.ts:158]: why are we still checking rule.vehicle_types is truthy here?

File: packages/mds-compliance-engine/engine/helpers.ts (hunk 3 above)
Comment (avatarneil) [reply packages/mds-compliance-engine/engine/helpers.ts:124]: Yeah, this should basically never throw, and was added for the sake of type safety without using an `as` assertion. The return type from `toFormat` is just a `string`, and because the template literal type `${number}:${number}:${number}` is not necessarily derivable from `string`, we need to either assert the type, or write a typeguard. I figured the extra sub-ms of runtime using the typeguard was nicer than an assertion, but totally up for changing it to an assertion because we're basically guaranteed that this conversion succeeds (unless there's a regression in the `luxon` library, which is unlikely).

File: packages/mds-compliance-engine/engine/helpers.ts [hunk 5]
```diff
@@ -1,17 +1,28 @@
...
 import { areThereCommonElements, isDefined, now, RuntimeError } from '@mds-core/mds-utils'
+import { DateTime } from 'luxon'
 import moment from 'moment-timezone'
 import { ProviderInputs, VehicleEventWithTelemetry } from '../@types'
 
 const { env } = process
 
 const TWO_DAYS_IN_MS = 172800000
 
+const isOfTimeFormat = (timeString: string): timeString is TIME_FORMAT => /^\d{2}:\d{2}:\d{2}$/.test(timeString)
```
[the remark this answers is not in the document]
Comment (avatarneil) [reply packages/mds-compliance-engine/engine/helpers.ts:24]: It's soooooo nice 😍

File: packages/mds-compliance-engine/engine/helpers.ts (hunk 4 above)
Comment (avatarneil) [reply packages/mds-compliance-engine/engine/helpers.ts:158]: Because the type of `vehicle_types` here is `?: Nullable<VEHICLE_TYPE[]>`, so it could be one of:
```
undefined | null | Array<VEHICLE_TYPE>
```

the `undefined` bit comes from the `?:` operator in the original typedef for this property, which can be found in packages/mds-policy-service/@types/models.ts#83

File: packages/mds-compliance-engine/engine/helpers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (EthanSherr) [reply packages/mds-compliance-engine/engine/helpers.ts:24]: woa!

**Review by EthanSherr — APPROVED** [review]
ran on city-dev and seems to be generating violations

File: packages/mds-compliance-engine/engine/helpers.ts (hunk 3 above)
Comment (janedotx) [reply packages/mds-compliance-engine/engine/helpers.ts:124]: Ah. Maybe leave a comment that this is essentially a type guard, then.
